src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostUserInterfaceSecurity.cs

```diff
@@ -100,25 +126,84 @@ class ConsoleHostUserInterface : System.Management.Automation.Host.PSHostUserInt
                 while (userName.Length == 0);
             }
 
-            passwordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_Password, userName
-            );
+            passwordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_Password, userName);
 
-            //
             // now, prompt for the password
-            //
-            WriteToConsole(passwordPrompt, true);
-            password = ReadLineAsSecureString();
-            if (password == null)
+            do
             {
-                return null;
+                WriteToConsole(passwordPrompt, true);
+                password = ReadLineAsSecureString();
+                if (password == null)
+                {
+                    return null;
+                }
             }
+            while (password.Length == 0);
 
-            WriteLineToConsole();
+            if (reenterPassword)
+            {
+                reenterPasswordPrompt = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_ReenterPassword, userName);
+                passwordMismatch = StringUtil.Format(ConsoleHostUserInterfaceSecurityResources.PromptForCredential_PasswordMismatch);
 
-            cred = new PSCredential(userName, password);
+                // now, prompt to re-enter the password.
+                WriteToConsole(reenterPasswordPrompt, true);
+                confirmPassword = ReadLineAsSecureString();
+                if (confirmPassword == null)
+                {
+                    return null;
+                }
 
+                if (!SecureStringEquals(password, confirmPassword))
+                {
+                    WriteToConsole(ConsoleColor.Red, ConsoleColor.Black, passwordMismatch, false);
+                    return null;
+                }
+            }
+
+            WriteLineToConsole();
+            cred = new PSCredential(userName, password);
             return cred;
         }
+
+        private static bool SecureStringEquals(SecureString password, SecureString confirmPassword)
+        {
+            if (password.Length != confirmPassword.Length)
+            {
+                return false;
+            }
+
+            IntPtr pwd_ptr = IntPtr.Zero;
+            IntPtr confirmPwd_ptr = IntPtr.Zero;
+            try
+            {
+                pwd_ptr = Marshal.SecureStringToBSTR(password);
+                confirmPwd_ptr = Marshal.SecureStringToBSTR(confirmPassword);
+
+                int pwdLength = Marshal.ReadInt32(pwd_ptr, -4);
+                int equal = 0;
+                for (int i = 0; i < pwdLength; i++)
+                {
+                    byte c1 = Marshal.ReadByte(pwd_ptr + i);
```

Comment:
Marshal.ReadByte(), can also take an offset parameter, instead of incrementing the pointer reference.  Not sure if it makes a difference.
```
Marshal.ReadByte(pwd_ptr, i);
```

Reply:
Looks more readable.